Adyen/UI/View Controllers/SearchViewController.swift

```diff
@@ -7,38 +7,73 @@
 import UIKit
 
 @_spi(AdyenInternal)
-public protocol SearchViewControllerDelegate: AnyObject {
-
-    func textDidChange(_ searchBar: UISearchBar, searchText: String)
+public protocol SearchViewControllerEmptyView: UIView {
+    var searchTerm: String { get set }
 }
 
 @_spi(AdyenInternal)
 public class SearchViewController: UIViewController {
 
-    internal let childViewController: UIViewController
+    public typealias ResultProvider = (_ searchTerm: String, _ handler: @escaping ([ListItem]) -> Void) -> Void
+    
+    internal enum InterfaceState {
+        case loading
+        case empty(searchTerm: String)
+        case showingResults(results: [ListItem])
+    }
+    
+    internal private(set) var interfaceState: InterfaceState = .empty(searchTerm: "") {
+        didSet {
+            guard interfaceState != oldValue else { return }
+            updateInterface()
+        }
+    }
+    
+    public lazy var resultsListViewController = ListViewController(style: style)
     private let localizationParameters: LocalizationParameters?
-    private let style: ViewStyle
-
-    public init(childViewController: UIViewController,
-                style: ViewStyle,
-                delegate: SearchViewControllerDelegate,
-                localizationParameters: LocalizationParameters? = nil) {
-        self.childViewController = childViewController
+    internal let style: ViewStyle
+    private let searchBarPlaceholder: String?
+    private let resultProvider: ResultProvider
+    private let shouldFocusSearchBarOnAppearance: Bool
+    
+    public init(
+        style: ViewStyle,
+        searchBarPlaceholder: String? = nil,
+        emptyView: SearchViewControllerEmptyView,
+        shouldFocusSearchBarOnAppearance: Bool = false,
+        localizationParameters: LocalizationParameters? = nil,
+        resultProvider: @escaping ResultProvider
+    ) {
+        self.emptyView = emptyView
         self.style = style
-        self.delegate = delegate
+        self.resultProvider = resultProvider
+        self.searchBarPlaceholder = searchBarPlaceholder ?? localizedString(.searchPlaceholder, localizationParameters)
         self.localizationParameters = localizationParameters
+        self.shouldFocusSearchBarOnAppearance = shouldFocusSearchBarOnAppearance
+        
         super.init(nibName: nil, bundle: nil)
     }
 
     @available(*, unavailable)
     internal required init?(coder: NSCoder) {
         fatalError("init(coder:) has not been implemented")
     }
+    
+    internal let emptyView: SearchViewControllerEmptyView
 
-    private lazy var searchBar: UISearchBar = {
+    internal lazy var loadingView: UIActivityIndicatorView = {
+        let loadingView = UIActivityIndicatorView(style: .whiteLarge)
+        if #available(iOS 13.0, *) {
+            loadingView.color = .secondarySystemFill
+        }
```

Comment:
Maybe add this colors to `Style`?
Are those colors darkmode-friendly?

Reply:
Great point regarding styling, will add that in a separate PR 👌. For now I changed it to use our default loading color.

Reply:
TBH, I would prefer to address it here. But up to you